File: domain/src/main/scala/net/kemitix/thorp/domain/LocalFile.scala
package net.kemitix.thorp.domain

import java.io.File

import scala.jdk.OptionConverters._

import net.kemitix.thorp.domain.HashType.MD5

final case class LocalFile private (
    file: File,
    source: File,
    hashes: Hashes,
    remoteKey: RemoteKey,
    length: Long
)

object LocalFile {
  val remoteKey: SimpleLens[LocalFile, RemoteKey] =
    SimpleLens[LocalFile, RemoteKey](_.remoteKey,
                                     b => a => b.copy(remoteKey = a))
  def matchesHash(localFile: LocalFile)(other: MD5Hash): Boolean =
    localFile.hashes.values.contains(other)
  def md5base64(localFile: LocalFile): Option[String] =
    localFile.hashes.get(MD5).toScala.map(h => h.hash64())
}

File: domain/src/main/java/net/kemitix/thorp/domain/LocalFile.java
package net.kemitix.thorp.domain;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.RequiredArgsConstructor;

import java.io.File;
import java.util.Optional;

@EqualsAndHashCode
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class LocalFile {
    public final File file;
    public final File source;
    public final Hashes hashes;
    public final RemoteKey remoteKey;
    public final Long length;
    public static LocalFile create(
            File file,
            File source,
            Hashes hashes,
            RemoteKey remoteKey,
            Long length
    ) {
        return new LocalFile(file, source, hashes, remoteKey, length);
    }
    public boolean matchesHash(MD5Hash hash) {
        return hashes.values().contains(hash);
    }
    public Optional<String> md5base64() {
        return hashes.get(HashType.MD5).map(MD5Hash::hash64);
    }
}
